From rientjes@google.com  Mon May 11 16:03:26 2009
From: David Rientjes <rientjes@google.com>
Date: Mon, 11 May 2009 15:45:14 -0700 (PDT)
Subject: Staging: android: lowmemorykiller: fix possible android low memory killer NULL pointer
To: Greg Kroah-Hartman <gregkh@suse.de>
Cc: Arve Hjønnevåg <arve@android.com>, San Mehat <san@android.com>
Message-ID: <alpine.DEB.2.00.0905111544140.5533@chino.kir.corp.google.com>

From: David Rientjes <rientjes@google.com>

get_mm_rss() atomically dereferences the actual without checking for a
NULL pointer, which is possible since task_lock() is not held.

Cc: San Mehat <san@android.com>
Cc: Arve Hjønnevåg <arve@android.com>
Signed-off-by: David Rientjes <rientjes@google.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/staging/android/lowmemorykiller.c |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

--- a/drivers/staging/android/lowmemorykiller.c
+++ b/drivers/staging/android/lowmemorykiller.c
@@ -92,12 +92,18 @@ static int lowmem_shrink(int nr_to_scan,
 	for_each_process(p) {
 		int oom_adj;
 
-		if (!p->mm)
+		task_lock(p);
+		if (!p->mm) {
+			task_unlock(p);
 			continue;
+		}
 		oom_adj = p->oomkilladj;
-		if (oom_adj < min_adj)
+		if (oom_adj < min_adj) {
+			task_unlock(p);
 			continue;
+		}
 		tasksize = get_mm_rss(p->mm);
+		task_unlock(p);
 		if (tasksize <= 0)
 			continue;
 		if (selected) {
